### DISCUSSION TITLE
Painting Foliage using GRIDMAP

### DISCUSSION THREAD

**Discussion by sangeethpp** [godotengine/godot-proposals#4079] [3D]
I think  GridMap can be used as a Foliage painting tool if it can detect the target mesh/terrain height automatically...Now we have to use Q/E keys which is not practical for uneven terrains and only one mesh is possible at a time...Godot already has great Mesh library, which can be created very easily .I know the orginal purpose of Gridmap is far from this ... But It'll be great If this feature can be added in Godot 4...

**Comment by wareya** [comment]
Not just foliage, I think being able to easily "paint" gridmap objects off-grid like this would be great for level designers in general.

**Reply by sangeethpp** [reply]
Yes...this will solve the scatter tool problems and useful for  artist who can't code....

**Comment by Calinou** [comment]
If GridMap objects are to be placed off-grid, then it's not really a GridMap anymore :slightly_smiling_face: 

Have you looked at <https://github.com/HungryProton/scatter>?

**Reply by sangeethpp** [reply]
Scatter add on is cool... But it doesn't support Mesh library... The way Gridmap works is great.. Just click and paint...and only one node in the project..

**Reply by sangeethpp** [reply]
OK...is there anyway to align Gridmap to the target terrain ? I mean.. can I bend the Gridmap..? 😀

**Reply by theraot** [reply]
@sangeethpp Have you tried MultiMeshInstance? I think you should consider this idea for MultiMeshInstance instead of GridMap. MultiMeshInstance will allow you to have multiple instances of the same mesh in a single Node in the scene, except currently you don't really have any freedom to place them from the editor (there is an option that appears in the toolbar to populate it). We certainly can do better by placing them from a tool script. Thus a more direct way to place it would be welcome.

**Reply by wareya** [reply]
Is it possible to load a mesh library into a MultiMeshInstance? Forgive me of I'm wrong but I don't think it really overlaps with GridMap in terms of design or workflow.

**Reply by sangeethpp** [reply]
Multimesh is very useful.. But it  definitely need some kinda weight paint/vertex paint support and  the Mesh Library also...a combination of Gridmap/Multimesh feature set  is what we need..

**Reply by theraot** [reply]
@wareya No, you cannot load a MeshLibrary in a MultiMeshInstance. Just a MultiMesh, which is a mesh repeated a bunch of times. However, It is possible to make something like a GridMap with a MultiMeshInstance for each Mesh you want to be able place... Except there isn't UI to do that. So, of course it does not overlap with it in terms of workflow, but I believe a paint feature can be implemented over MultiMeshInstance much easier, in terms of the code behind the scenes, I mean.

@sangeethpp GridMap is overdue for a rework. However, it is meant to be on a gird. It is the TileMap for 3D. It does not store a transform for each thing you place, instead it stores an orientation index which is one of 24 possible values, and it only stores things at integer positions (cells), and its optimizations are based around that. Sadly, as I was saying MultiMeshInstance lacks UI to place the meshes. Currently the UI can only place them around another existing mesh. I can imagine adding such tooling, which could be done with an addon, and... Well, as @Calinou points out there are addons already.

So, perhaps something that mixes GridMap and MultiMeshInstance makes sense.

---

By the way, my take on improving GridMap is that it should not need a Scene to make MeshLibrary. Instead there should be a MeshLibraby editor, the same way there is a TileSet editor. Having a Scene we can modify, where we can add things that the MeshLibrary does not support, and having to convert it again after the changes… Is there something to like in that workflow? I would like to see a proposal to improve GridMap usability, in particular that workflow. However, I would also like to see a proposal to improve MultiMeshInstance usability, in particular the way the instances are placed.

Furthermore, I think this paint feature fits better MultiMeshInstance since it is not on a grid, and MultiMeshInstance is meant to have arbitrary orientations and positions and large numbers. *Oh, that reminds me, you can also do the scattering with particles. Hint: have the particles not move.*

Now, let us find the common ground. So you want a MultiMeshInstance with MeshLibrary? What if, we add a MultiMultiMeshInstance (for lack of a better name) that allows you to load MeshLibrary, and what it does is create a bunch of MultiMeshInstance behind the scenes? Then it would be similar to a GridMap, without the Grid… a …Map. I don't know, I'm just tossing the idea out there.

**Reply by sangeethpp** [reply]
OK... For less confusion let us don't touch the Gridmap.. Let it be there like as it is...But can we have a tool in Godot 4 that can paint foliage on a terrain...? with Mesh Library...?
@theraot  thank you...

**Reply by sangeethpp** [reply]
Hi ...I've uploaded a video just to show what I was talking about....pls check..,thank you...
https://youtu.be/6gAg3NnZjA8

**Reply by sangeethpp** [reply]
I think I failed to explain it properly...I meant the usage of built in mesh library for painting the foliage... I used gridmap to show that as I had no other options... And I thought tweaking the Grid map will solve everything...
